design/phase-level-progress-tracking.md

```diff
@@ -0,0 +1,143 @@
+ type Phase struct {
+        Name     string                 `json:"name"`
+        State    State                  `json:"state"`
+        Output   map[string]interface{} `json:"output,omitempty"`
++       Progress PhaseProgress          `json:"progress,omitempty"` <-----
+ }
+
++type PhaseProgress struct {
++       ProgressPercent       int64
++       LastTransitionTime    *metav1.Time
++       SizeUploadedB         int64
++       EstimatedTimeSeconds  int64
++       EstimatedUploadSizeB  int64
++}
+```
+
+### Action progress tracking
+
+Action can consists of multiple phases. These all phases can take consume different duration for completion based on the operations it performs. All phases for the action can be given equal weight-age and to calculate action progress completion %, average of all phase completion %ge can be calculated
+
+```
+% completion of action = sum(% Phase completion of all the phases)/ no. of phases
+```
+
+### Phase progress tracking
+
+Each function is designed to perform different operations in different ways. Since the operation execution is specific to function, each function can be different ways to calculate and provide info about the progress. E.g datamover functions can leverage kopia stats to track progress. So it makes sense to delegate progress tracking to function implementation.
+
+Each Kanister function implements Progress interface. The Kanister function implementation calculates and returns the progress stats.
+
+**Progress interface**
+
+```
+type Progress interface {
+        Stats() (*FuncProgress, error)
+}
+
+type FuncProgress struct {
+        ProgressPercent      int64
+        SizeUploadedB        int64
+        EstimatedTimeSeconds int64
+        EstimatedUploadSizeB int64
+}
+```
+
+**Kanister function interface**
+
+```
+type Func interface {
++       Progress
+        RequiredArgs() []string
+        Arguments() []string
+        Exec(context.Context, param.TemplateParams, map[string]interface{}) (map[string]interface{}, error)
+ }
+```
+
+The PhaseProgress.ProgressPercent would be the of progress the functions it consists of.
+
+### Kanister functions progress tracking
+
+On high level we can divide the Kanister functions into two groups
+
+### Datamover Kanister functions
+
+Kanister data mover functions (like BackupData, CopyVolumeData, RestoreData, etc) use Kopia to snapshot filesystem and move data to/from objects.
+
+While snapshotting, kopia provides the info about progress on the terminal
+
+```
+$ kopia --log-level=error --config-file=/tmp/kopia-repository.config --log-dir=/tmp/kopia-log snapshot --progress-update-interval=5s /data
+ / 1 hashing, 118 hashed (546.9 MB), 0 cached (0 B), uploaded 0 B, estimated 4.5 GB (12.1%) 35s left
+ * 0 hashing, 136 hashed (1.4 GB), 0 cached (0 B), uploaded 244 B, estimated 4.5 GB (30.1%) 30s left
+```
+
+The command output can be parsed to get the progress metadata.
+
+### Non-datamover Kanister functions
+
+TBD
```

Comment:
Even though it's not decided as of now, to actually implement this we will have to have something here right? For example either hard code or something else, otherwise would we be able to implement this design?

Reply:
Makes sense. We have discussed a couple of idea, will add it here. Thanks